layer_metadata_restorer.go

```diff
@@ -45,22 +50,60 @@ func (la *DefaultLayerMetadataRestorer) Restore(buildpacks []buildpack.GroupBuil
 func (la *DefaultLayerMetadataRestorer) restoreStoreTOML(appMeta platform.LayersMetadata, buildpacks []buildpack.GroupBuildpack) error {
 	for _, bp := range buildpacks {
 		if store := appMeta.MetadataForBuildpack(bp.ID).Store; store != nil {
-			if err := WriteTOML(filepath.Join(la.LayersDir, launch.EscapeID(bp.ID), "store.toml"), store); err != nil {
+			if err := WriteTOML(filepath.Join(la.layersDir, launch.EscapeID(bp.ID), "store.toml"), store); err != nil {
 				return err
 			}
 		}
 	}
 	return nil
 }
 
+type layerToSha struct {
+	layerToShaMap map[string]string
+}
+
+func initializeLayerToSha() layerToSha {
+	return layerToSha{make(map[string]string)}
+}
+
+type buildpackLayersToSha struct {
+	buildpacksToLayersShaMap map[string]layerToSha
+}
+
+func initializeBuildpackLayersToSha() buildpackLayersToSha {
+	return buildpackLayersToSha{make(map[string]layerToSha)}
+}
+
+// if the layer exists for the buildpack ID, its SHA will be returned. Otherwise, an empty string will be returned.
+func (bls *buildpackLayersToSha) getShaByBuildpackLayers(buildpackID, layerName string) string {
+	if layerToSha, buildpackExists := bls.buildpacksToLayersShaMap[buildpackID]; buildpackExists {
+		if sha, layerExists := layerToSha.layerToShaMap[layerName]; layerExists {
+			return sha
+		}
+	}
+	return ""
+}
+
+func (bls *buildpackLayersToSha) addLayerToMap(buildpackID, layerName, sha string) {
+	_, exists := bls.buildpacksToLayersShaMap[buildpackID]
+	if !exists {
+		bls.buildpacksToLayersShaMap[buildpackID] = initializeLayerToSha()
+	}
+	bls.buildpacksToLayersShaMap[buildpackID].layerToShaMap[layerName] = sha
+}
+
 func (la *DefaultLayerMetadataRestorer) restoreLayerMetadata(appMeta platform.LayersMetadata, cacheMeta platform.CacheMetadata, buildpacks []buildpack.GroupBuildpack) error {
-	if la.SkipLayers {
-		la.Logger.Infof("Skipping buildpack layer analysis")
+	if la.skipLayers {
+		la.logger.Infof("Skipping buildpack layer analysis")
 		return nil
 	}
 
+	if !la.useShaFile {
+		la.buildpacksToLayersSha = initializeBuildpackLayersToSha()
+	}
```

Comment:
We can but I don't like it because we call `NewLayerMetadataRestorer` even for `platform API < 0.7` and then I thought that maybe it's better that the map won't be initialized at all.
I don't have strong opinion on it, just didn't want to touch it if not necessary.

Reply:
But `NewLayerMetadataRestorer` has access to `useShaFile`, no?

Reply:
You're right. I forgot that I added it as part of one of the commits that I made yesterday.
I can move the `initializeBuildpackLayersToSha` there. Thanks.

Reply:
I moved it to `NewLayerSHAStore`. Please see my coming commit.